documentation/specs/dotnet-run-for-maui.md

```diff
@@ -0,0 +1,222 @@
+
+* Each platform has different behavior, regards to displaying console
+  output, etc.
+
+* Using an MSIX with WindowsAppSDK doesn't support `dotnet run` at all,
+  relying completely on IDEs like Visual Studio.
+
+* `dotnet run` does not have a concept of a "deploy" step.
+
+This has become more relevant in the AI era, as someone is going to
+expect AIs in "agent mode" to build and run their app. If the
+command-line options are difficult, AIs will fail just as badly as
+humans do today.
+
+## The `dotnet run` Pipeline
+
+These are the high-level steps during `dotnet run`, that we would like
+to make extensible for .NET MAUI (and future) scenarios.
+
+* `restore`: unchanged
+
+* "Pre-run evaluation"
+
+  * If the project is multi-targeted, containing the
+    `$(TargetFrameworks)` property _and_ that property has more than
+    one item in it, and `-f` was not supplied...
+
+    * Prompt the user to select from a list of the
+      `$(TargetFrameworks)`
+
+    * Non-interactive mode will give a friendly error message,
+      suggesting to supply the `-f` property, listing available target
+      frameworks in the project.
+
+  * Once a `$(TargetFramework)` is selected, either from previous
+    steps or `-f`...
+
+    * If a `ComputeAvailableDevices` MSBuild target is available, provided by
+      the iOS or Android workload, etc. ...
+
+    * Call the MSBuild target, which returns a list of `@(Devices)` items...
+
+```xml
+<ItemGroup>
+  <!-- Android examples -->
+  <Devices Include="emulator-5554"  Description="Pixel 7 - API 35" Type="Emulator" Status="Offline" />
+  <Devices Include="emulator-5555"  Description="Pixel 7 - API 36" Type="Emulator" Status="Online" />
+  <Devices Include="0A041FDD400327" Description="Pixel 7 Pro"      Type="Device"   Status="Online" />
+  <!-- iOS examples -->
+  <Devices Include="94E71AE5-8040-4DB2-8A9C-6CD24EF4E7DE" Description="iPhone 11 - iOS 18.6" Type="Simulator" Status="Shutdown" />
+  <Devices Include="FBF5DCE8-EE2B-4215-8118-3A2190DE1AD7" Description="iPhone 14 - iOS 26.0" Type="Simulator" Status="Booted" />
+  <Devices Include="23261B78-1E31-469C-A46E-1776D386EFD8" Description="My iPhone 13"         Type="Device"    Status="Unavailable" />
+  <Devices Include="AF40CC64-2CDB-5F16-9651-86BCDF380881" Description="My iPhone 15"         Type="Device"    Status="Paired" />
+</ItemGroup>
+```
+
+_NOTE: each workload can decide which metadata values for `%(Type)`
+and `%(Status)` are useful, filtering offline devices, etc. The output
+above would be analogous to running `adb devices`, `xcrun simctl list
+devices`, or `xcrun devicectl list devices`._
+
+* Continuing on...
+
+  * Prompt the user to select from this list of devices, emulators,
+    or simulators.
+
+  * Non-interactive mode will error, suggesting to supply the
+    `--device` switch. Listing the options returned by the
+    `ComputeAvailableDevices` MSBuild target.
+
+* `build`: unchanged, but is passed `-p:Device`.
```

Comment:
Oh I see the build step is mentioned here already. But doesn't it happen *before* the "pre-run evaluation"?

Reply:
An iOS project is built differently depending on whether you're building for simulator or device.

This means that you can either first select to run for the simulator or device, and then we'll present all the simulators or all the devices for you to choose from, or we can first present all the simulators + all the devices, and then we automatically build correctly depending on whether you chose a simulator or a device.